.Net application aborts at startup with "Failed to create CoreCLR, HRESULT: 0x80004005" - .Net 7 preview 5

### Description

I have a project here: https://github.com/Webreaper/Damselfly which I upgraded to .Net 7 preview 5 this evening, and also updated EF Core and a bunch of other dependencies (https://github.com/Webreaper/Damselfly/compare/master...develop). 

The app compiles and runs fine on my M1 Macbook (I'm using VS for Mac 2022). However, when built and deployed to a docker container, and launched on Linux (I'm using a Synology NAS) it fails with:

`Failed to create CoreCLR, HRESULT: 0x80004005`

The same app launched fine built and ran (on both OSX and Linux) against .Net 7 preview 4. 

I'll do some additional tests tomorrow to see if I can narrow down to the specific update that cause this to break (i.e., is it .Net 7 preview 5 itself, or some other dep, or a combination of the two). The app is a Blazor Server app running as a self-contained single-file executable, but since it never gets as far as initialising the CLR I don't think that Blazor or ASP.Net is relevant. I've also tried disabling trimming, but it didn't resolve the issue.

I guess my first question is - how do I go about getting additional debugging information to figure out what's causing this problem?

### Reproduction Steps

Take my app, build and run it against .Net 7 preview 4. It works. Upgrade to preview 5, it fails to start. 

### Expected behavior

App runs successfully on both OSX and Linux.

### Actual behavior

App runs successfully on OSX, but on Linux the app fails to start with `Failed to create CoreCLR, HRESULT: 0x80004005`

### Regression?

Yes, this app worked with .Net 7 preview 4.

### Known Workarounds

_No response_

### Configuration

_No response_

### Other information

_No response_

Hey @am11, I just tried that, but it's failing to pull packages from Nuget in my Github build, e.g.:

```
/home/runner/work/Damselfly/Damselfly/Damselfly.Migrations.Sqlite/Damselfly.Migrations.Sqlite.csproj : error NU1102: Unable to find package Microsoft.NETCore.App.Runtime.linux-x64 with version (= 7.0.0-preview.6.22312.1) [/home/runner/work/Damselfly/Damselfly/Damselfly.Web/Damselfly.Web.csproj]
/home/runner/work/Damselfly/Damselfly/Damselfly.Migrations.Sqlite/Damselfly.Migrations.Sqlite.csproj : error NU1[10](https://github.com/Webreaper/Damselfly/runs/6895291025?check_suite_focus=true#step:5:11)2:   - Found 89 version(s) in nuget.org [ Nearest version: 7.0.0-preview.5.22301.12 ] [/home/runner/work/Damselfly/Damselfly/Damselfly.Web/Damselfly.Web.csproj]
```
Is there anything else I need to switch to allow me to pull that newer version (I've not used the dailies before)?

--- a/src/coreclr/minipal/Unix/doublemapping.cpp
+++ b/src/coreclr/minipal/Unix/doublemapping.cpp
@@ -23,6 +23,18 @@
 
 #if defined(TARGET_OSX) && defined(TARGET_AMD64)
 #include <mach/mach.h>
+#include <sys/sysctl.h>
+
+bool IsProcessTranslated()
+{
+   int ret = 0;
+   size_t size = sizeof(ret);
+   if (sysctlbyname("sysctl.proc_translated", &ret, &size, NULL, 0) == -1)
+   {
+      return false;
+   }
+   return ret != 0;
+}
 #endif // TARGET_OSX && TARGET_AMD64
 
 #ifndef TARGET_OSX
@@ -65,6 +77,15 @@ bool VMToOSInterface::CreateDoubleMemoryMapper(void** pHandle, size_t *pMaxExecu
     *pMaxExecutableCodeSize = MaxDoubleMappedSize;
     *pHandle = (void*)(size_t)fd;
 #else // !TARGET_OSX
+
+#ifdef TARGET_AMD64
+    if (IsProcessTranslated())
+    {
+        // Rosetta doesn't support double mapping correctly
+        return false;
+    }
+#endif // TARGET_AMD64
+
     *pMaxExecutableCodeSize = SIZE_MAX;
     *pHandle = NULL;
 #endif // !TARGET_OSX

--- a/src/coreclr/minipal/Unix/doublemapping.cpp
+++ b/src/coreclr/minipal/Unix/doublemapping.cpp
@@ -33,7 +33,7 @@ bool IsProcessTranslated()
    {
       return false;
    }
-   return ret != 0;
+   return ret == 1;
 }
 #endif // TARGET_OSX && TARGET_AMD64
 

--- a/src/coreclr/utilcode/executableallocator.cpp
+++ b/src/coreclr/utilcode/executableallocator.cpp
@@ -249,7 +249,8 @@ bool ExecutableAllocator::Initialize()
     {
         if (!VMToOSInterface::CreateDoubleMemoryMapper(&m_doubleMemoryMapperHandle, &m_maxExecutableCodeSize))
         {
-            return false;
+            g_isWXorXEnabled = false;
+            return true;
         }
 
         m_CriticalSection = ClrCreateCriticalSection(CrstExecutableAllocatorLock,CrstFlags(CRST_UNSAFE_ANYMODE | CRST_DEBUGGER_THREAD));

